zanata-client-commands/src/main/java/org/zanata/client/commands/PushPullCommand.java

```diff
@@ -245,4 +257,108 @@ protected void storeETagCache() {
+                // TODO server statistics API should return locale with alias
+                TranslatedPercent translatedPercent =
+                        new TranslatedPercent(statsForSingleLocale.getTotal(),
+                                statsForSingleLocale.getTranslatedOnly(),
+                                statsForSingleLocale.getApproved());
+
+                localeToStatsBuilder.put(
+                        new LocaleId(statsForSingleLocale.getLocale()),
+                        translatedPercent);
+            }
+            Map<LocaleId, TranslatedPercent> localeStats =
+                    localeToStatsBuilder.build();
+            docIdToStatsBuilder.put(docId, localeStats);
+        }
+        return docIdToStatsBuilder.build();
+    }
+
+    @VisibleForTesting
+    protected ContainerTranslationStatistics getDetailStatisticsForProjectVersion() {
+        return statsClient
+                    .getStatistics(getOpts().getProj(),
+                            getOpts().getProjectVersion(), true, false, null);
+    }
+
+    /**
+     * this stats map will have docId as key, the value is another map with
+     * localeId as key and translated percent as value.
+     * It's optional if we require statistics to determine which file to pull.
+     * In cases where statistics is not required,
+     * i.e. pull source only or minimum percent is set to 0, this will be
+     * Optional.absence().
+     *
+     * @param pullTarget whether we need to pull translation target
+     * @return either detailed document statistics or optional.absence()
+     */
+    protected Optional<Map<String, Map<LocaleId, TranslatedPercent>>> prepareStatsIfApplicable(
+            boolean pullTarget) {
+
+        Optional<Map<String, Map<LocaleId, TranslatedPercent>>> optionalStats =
+                Optional.absent();
+        O opts = getOpts();
+        if (pullTarget && opts instanceof PullOptions &&
+                ((PullOptions) opts).getMinDocPercent() > 0) {
+            optionalStats = Optional.of(getDocsTranslatedPercent());
+        }
+        return optionalStats;
+    }
+
+    protected boolean shouldPullThisLocale(
+            Optional<Map<String, Map<LocaleId, TranslatedPercent>>> optionalStats,
+            String localDocName, LocaleId serverLocale) {
+        int minDocPercent = ((PullOptions) getOpts()).getMinDocPercent();
+        if (log.isDebugEnabled() && optionalStats.isPresent()) {
+            log.debug("{} for locale {} is translated {}%", localDocName,
+                    serverLocale, optionalStats.get()
+                            .get(localDocName).get(serverLocale)
+                    .getTranslatedPercent());
+        }
+        return !optionalStats.isPresent()
+                || optionalStats.get().get(localDocName).get(serverLocale)
+                        .isAboveThreshold(minDocPercent);
+    }
+
+    protected static class TranslatedPercent {
+        protected final int translatedPercent;
```

Comment:
Why protected?

If you change int to double, it should be possible to get far better precision in the percentage checks.